### ISSUE TITLE
Error of calculating x_mask in the forward function of the class Conv2dSubsampling

### ISSUE THREAD

**Issue by Sxjdwang** [espnet/espnet#3037]
**Describe the bug**
The class Conv2dSubsampling (from nets.pytorch_backend.transformer.subsampling) is a layer block to conduct convolutional 2D subsampling (to 1/4 length) with 2 2D convolution (width=3, stride=2). Meanwhile, it provides the corresponding x mask of the output tensor. However, the code deriving the output x_mask is inaccuracy.

x_mask = x_mask[:, :, :-2:2][:, :, :-2:2] where the x_mask in the right hand side is the x mask of the input tensor.

For example, we have an input pair 
x ( in the shape of 2*10*83, where 10 is the number of frame and 83 is the number of feature dimensions) 
and x_mask (True, True, True, True, True, True, False, False, False, False, False;
                     True, True, True, True, True, True, True, True, True, True, True ). 
After the forward function, the x_mask should be (True, False; True, True). But the code outputs (True, True;True, True).

I propose a solution as below.

from espnet.nets.pytorch_backend.nets_utils import make_non_pad_mask

non_pad_len = x_mask.clone().long().sum(dim=(1, 2))
non_pad_len = ((non_pad_len - 3) // 2 - 2) // 2 +1
x_mask = x_mask[:, :, :-2:2][:, :, :-2:2]
x_mask = x_mask * make_non_pad_mask(non_pad_len.tolist()).to(x_mask.device).unsqueeze(-2)

Please correct me if I am wrong. Thanks.

**Comment by sw005320** [comment]
Thanks for your report.
I don't fully understand it, but if so, this would be a serious issue.
@ShigekiKarita, could you take a look at it?

**Comment by sw005320** [comment]
@Sxjdwang, did you observe some improvements with your fix?

**Comment by Sxjdwang** [comment]
@sw005320 I have not compared the performance of an ASR system with and without my fix. However,  I feed 83D-feature (80 filter bank and 3 pitch feature) of a piece of speech and the 83D-feature with some zero padding in the frame dimension to a transformer network with a Conv2dSubsampling layer. The latent features of them after the transformer encoder are different.

**Comment by Sxjdwang** [comment]
Sorry. The latent features of them after the transformer encoder are different without my fix.

**Comment by sw005320** [comment]
@pengchengguo will investigate this issue.

**Comment by pengchengguo** [comment]
Hi @Sxjdwang,

Thanks so much for your report.

I think this problem (I am not sure is an error) is caused by the difference of convolutional operation in PyTorch and subsampling operation in NumPy.

Assume we have two utterances in a batch, `x1.length = 15, x2.length = 11`, so `x2` will be padded to the same length as `x1`. The `x2` can be described as the diagram, where `T = True` means real frames and `F = False` means padding frames.

In the `Conv2dSubsampling` layer, we have two `Conv2d` layers with the same kernel size 3 and the same stride 2. So the length of subsampled output should be `((15 - 1) // 2 - 1) // 2 = 3`. To compute the subsampled mask quickly, we use `x_mask[:, :, :-2:2][:, :, :-2:2]` to refer to the convolution, which means we take the `T` or `F` flags every two-step. (So the flags of the 2nd line are the same as the red block, while the flags of the 1st line are the same as the blue block).

However, from our intuition, the real length of `x2` is 11, so it's output length should be `((11 - 1) // 2 - 1) // 2 = 2`, which means the mask of should be `[T, T, F]`.


<img width="668" alt="image" src="https://user-images.githubusercontent.com/23135280/111032201-04521f00-8446-11eb-8c24-e19465806508.png">

The receptive field of an output frame is 7, so if the 1st frame of its receptive field is `True`, this output frame will be marked as `True`. The inaccurate mask may include at most one padded frame into the computation. I think it may not influence the performance much.

**Comment by pengchengguo** [comment]
A most simple solution to solve this problem is generating the mask after subsampling. For example:

```
ilens_new = ((ilens - 1) // 2 - 1) // 2
hs_mask = make_non_pad_mask(ilens_new.tolist())
```

But I'm not sure whether such modifications will bring compatibility issues with our pre-trained models.

**Comment by pengchengguo** [comment]
Please correct me if I misunderstood.

**Comment by Sxjdwang** [comment]
@pengchengguo I agree with you that it may not influence the performance much although the original code generates inaccurate masks. Even it may enhance models' robustness.

**Comment by sw005320** [comment]
> @pengchengguo I agree with you that it may not influence the performance much although the original code generates inaccurate masks. Even it may enhance models' robustness.

I fully agree.
We'll revisit it after Interspeech.
My only concern is that we may not use our pre-trained model, but probably not. 
We can also keep some compatibility.
